test/e2e/util.go

```diff
@@ -319,5 +360,150 @@ func CreateSource(t *testing.T, clients *test.Clients, name string) context.Canc
+		clients.KubeClient.CoreV1().Services(simService.Namespace).Delete(context.Background(), simService.Name, metav1.DeleteOptions{})
+		clients.KubeClient.CoreV1().Secrets(simSecret.Namespace).Delete(context.Background(), simSecret.Name, metav1.DeleteOptions{})
+	}, t.Logf)
+	secret, err := clients.KubeClient.CoreV1().Secrets(simSecret.Namespace).Create(context.Background(), simSecret, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Error creating Secret: %v", err)
+	}
+	deployment, err := clients.KubeClient.AppsV1().Deployments(simDeployment.Namespace).Create(context.Background(), simDeployment, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Error creating VCSIM Deployment: %v", err)
+	}
+	service, err := clients.KubeClient.CoreV1().Services(simService.Namespace).Create(context.Background(), simService, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Error creating VCSIM Service: %v", err)
+	}
+
+	cancel := func() {
+		err := clients.KubeClient.AppsV1().Deployments(deployment.Namespace).Delete(context.Background(), deployment.Name, metav1.DeleteOptions{})
+		if err != nil {
+			t.Errorf("Error cleaning up Deployment %s", deployment.Name)
+		}
+		err = clients.KubeClient.CoreV1().Services(service.Namespace).Delete(context.Background(), service.Name, metav1.DeleteOptions{})
+		if err != nil {
+			t.Errorf("Error cleaning up Service %s", service.Name)
+		}
+		err = clients.KubeClient.CoreV1().Secrets(secret.Namespace).Delete(context.Background(), secret.Name, metav1.DeleteOptions{})
+		if err != nil {
+			t.Errorf("Error cleaning up Secret %s", secret.Name)
+		}
+	}
+
+	waitErr := wait.PollImmediate(test.PollInterval, test.PollTimeout, func() (bool, error) {
+		depl, err := clients.KubeClient.AppsV1().Deployments(ns).Get(context.Background(), simDeployment.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+
+		status := depl.Status
+		for i := range status.Conditions {
+			c := status.Conditions[i]
+			if c.Type == appsv1.DeploymentAvailable && c.Status == corev1.ConditionTrue {
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	if waitErr != nil {
+		cancel()
+		t.Fatalf("Error waiting for VCSIM deployment to be ready: %v", waitErr)
+	}
+
+	t.Log("vcsim ready")
+
+	return cancel
+}
+
+func newSimulator(namespace string) (*appsv1.Deployment, *corev1.Service) {
+	l := map[string]string{
+		"app": vcsim,
+	}
+
+	sim := appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      vcsim,
+			Namespace: namespace,
+			Labels:    l,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: pointer.Int32(1),
+			Selector: &metav1.LabelSelector{
+				MatchLabels: l,
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: l,
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{{
+						Name:  vcsim,
+						Image: "vmware/vcsim:latest",
```

Comment:
non-blocking: if we see `docker.io` `429`s we can always fall back to using `ko:` to build `vcsim` since it's already a vendored dependency.